In [1]:
import pandas as pd
import numpy as np
import dask.dataframe as dd
from pathlib import Path

In [2]:
path = Path('./data/')

In [3]:
list(path.iterdir())

[PosixPath('data/train_labels.csv'),
 PosixPath('data/test.csv'),
 PosixPath('data/specs.csv'),
 PosixPath('data/train.csv'),
 PosixPath('data/data-science-bowl-2019.zip'),
 PosixPath('data/sample_submission.csv')]

Read in train and test data

In [4]:
train_data = pd.read_csv(path/'train.csv')
train_labels = pd.read_csv(path/'train_labels.csv')
test_data = pd.read_csv(path/'test.csv')
specs = pd.read_csv(path/'specs.csv')
samp_submission = pd.read_csv(path/'sample_submission.csv')

In [11]:
train_data['id'] = list(range(train_data.shape[0]))
train_data['timestamp'] = pd.to_datetime(train_data.timestamp)

In [5]:
train_data.head()

event_id      game_session                 timestamp  \
0  27253bdc  45bb1e1b6b50c07b  2019-09-06T17:53:46.937Z   
1  27253bdc  17eeb7f223665f53  2019-09-06T17:54:17.519Z   
2  77261ab5  0848ef14a8dc6892  2019-09-06T17:54:56.302Z   
3  b2dba42b  0848ef14a8dc6892  2019-09-06T17:54:56.387Z   
4  1bb5fbdb  0848ef14a8dc6892  2019-09-06T17:55:03.253Z   

                                          event_data installation_id  \
0             {"event_code": 2000, "event_count": 1}        0001e90f   
1             {"event_code": 2000, "event_count": 1}        0001e90f   
2  {"version":"1.0","event_count":1,"game_time":0...        0001e90f   
3  {"description":"Let's build a sandcastle! Firs...        0001e90f   
4  {"description":"Let's build a sandcastle! Firs...        0001e90f   

   event_count  event_code  game_time                          title  \
0            1        2000          0        Welcome to Lost Lagoon!   
1            1        2000          0           Magma Peak - Level 1   
2            1        2000          0  Sandcastle Builder (Activity)   
3            2        3010         53  Sandcastle Builder (Activity)   
4            3        3110       6972  Sandcastle Builder (Activity)   

       type      world  
0      Clip       NONE  
1      Clip  MAGMAPEAK  
2  Activity  MAGMAPEAK  
3  Activity  MAGMAPEAK  
4  Activity  MAGMAPEAK

In [6]:
train_data.head()

event_id      game_session                 timestamp  \
0  27253bdc  45bb1e1b6b50c07b  2019-09-06T17:53:46.937Z   
1  27253bdc  17eeb7f223665f53  2019-09-06T17:54:17.519Z   
2  77261ab5  0848ef14a8dc6892  2019-09-06T17:54:56.302Z   
3  b2dba42b  0848ef14a8dc6892  2019-09-06T17:54:56.387Z   
4  1bb5fbdb  0848ef14a8dc6892  2019-09-06T17:55:03.253Z   

                                          event_data installation_id  \
0             {"event_code": 2000, "event_count": 1}        0001e90f   
1             {"event_code": 2000, "event_count": 1}        0001e90f   
2  {"version":"1.0","event_count":1,"game_time":0...        0001e90f   
3  {"description":"Let's build a sandcastle! Firs...        0001e90f   
4  {"description":"Let's build a sandcastle! Firs...        0001e90f   

   event_count  event_code  game_time                          title  \
0            1        2000          0        Welcome to Lost Lagoon!   
1            1        2000          0           Magma Peak - Level 1   
2            1        2000          0  Sandcastle Builder (Activity)   
3            2        3010         53  Sandcastle Builder (Activity)   
4            3        3110       6972  Sandcastle Builder (Activity)   

       type      world  
0      Clip       NONE  
1      Clip  MAGMAPEAK  
2  Activity  MAGMAPEAK  
3  Activity  MAGMAPEAK  
4  Activity  MAGMAPEAK

In [7]:
train_labels.head()

game_session installation_id                         title  \
0  6bdf9623adc94d89        0006a69f  Mushroom Sorter (Assessment)   
1  77b8ee947eb84b4e        0006a69f    Bird Measurer (Assessment)   
2  901acc108f55a5a1        0006a69f  Mushroom Sorter (Assessment)   
3  9501794defd84e4d        0006a69f  Mushroom Sorter (Assessment)   
4  a9ef3ecb3d1acc6a        0006a69f    Bird Measurer (Assessment)   

   num_correct  num_incorrect  accuracy  accuracy_group  
0            1              0       1.0               3  
1            0             11       0.0               0  
2            1              0       1.0               3  
3            1              1       0.5               2  
4            1              0       1.0               3

In [8]:
test_data.head()

event_id      game_session                 timestamp  \
0  27253bdc  0ea9ecc81a565215  2019-09-10T16:50:24.910Z   
1  27253bdc  c1ea43d8b8261d27  2019-09-10T16:50:55.503Z   
2  27253bdc  7ed86c6b72e725e2  2019-09-10T16:51:51.805Z   
3  27253bdc  7e516ace50e7fe67  2019-09-10T16:53:12.825Z   
4  7d093bf9  a022c3f60ba547e7  2019-09-10T16:54:12.115Z   

                                          event_data installation_id  \
0             {"event_code": 2000, "event_count": 1}        00abaee7   
1             {"event_code": 2000, "event_count": 1}        00abaee7   
2             {"event_code": 2000, "event_count": 1}        00abaee7   
3             {"event_code": 2000, "event_count": 1}        00abaee7   
4  {"version":"1.0","round":0,"event_count":1,"ga...        00abaee7   

   event_count  event_code  game_time                    title  type  \
0            1        2000          0  Welcome to Lost Lagoon!  Clip   
1            1        2000          0     Magma Peak - Level 1  Clip   
2            1        2000          0     Magma Peak - Level 2  Clip   
3            1        2000          0  Crystal Caves - Level 1  Clip   
4            1        2000          0                Chow Time  Game   

          world  
0          NONE  
1     MAGMAPEAK  
2     MAGMAPEAK  
3  CRYSTALCAVES  
4  CRYSTALCAVES

In [9]:
specs.tail()

event_id                                               info  \
381  29f54413  The start round event is triggered at the star...   
382  06372577  The start tutorial event is triggered at the s...   
383  2a444e03  This event occurs when the player picks up a w...   
384  9e6b7fb5  This event occurs when the player clicks on th...   
385  d3640339  This event occurs when the player clicks on th...   

                                                  args  
381  [{"name":"game_time","type":"int","info":"mill...  
382  [{"name":"game_time","type":"int","info":"mill...  
383  [{"name":"game_time","type":"int","info":"mill...  
384  [{"name":"game_time","type":"int","info":"mill...  
385  [{"name":"game_time","type":"int","info":"mill...

In [10]:
samp_submission.head()

installation_id  accuracy_group
0        00abaee7               3
1        01242218               3
2        017c5718               3
3        01a44906               3
4        01bc6cb6               3

# Exploratory analysis

## How many unique events are there?

In [60]:
specs.event_id.nunique()

386

A total of 386 unique events

## What are the most common events?

In [71]:
event_freq = train_data.groupby('event_id',as_index=False) \
                        .agg({'timestamp':'count'}) \
                        .sort_values('timestamp',ascending=False) \
                        .merge(specs,on='event_id')
for idx,row in event_freq.iterrows():
    print(row['timestamp'])
    print(row['info'])
    print()

274673
This event occurs when the player clicks on something that isn’t covered elsewhere. It can be useful in determining if there are attractive distractions (things the player think should do something, but don’t) in the game, or diagnosing players who are having mechanical difficulties (near misses).

256179
This event occurs when the player picks up a bottle by clicking on it, or begins dragging a bottle. It contains information about the bottle that was picked up and the position of the click.

224694
This event occurs when the player clicks on something that isn't covered elsewhere. It can be useful in determining if there are attractive distractions (things the player think should do something, but don't) in the game, or diagnosing players who are having mechanical difficulties (near misses).

206129
This event occurs when the player picks up a bug by clicking on it, or begins dragging a bug. It contains information about the bug that was picked up and the position of the click

1074
This event occurs when the player clicks on the help button. It contains information about the state of the game that caused the player to seek help. This event is usually followed by a 3010 (system-initiated instruction) event. It helps identify points that players feel are too difficult.

846
This event occurs when the player clicks on the help button. It contains information about the state of the game that caused the player to seek help. This event is usually followed by a 3010 (system-initiated instruction) event. It helps identify points that players feel are too difficult.

775
This event occurs when the player clicks on the help button. It contains information about the state of the game that caused the player to seek help. This event is usually followed by a 3010 (system-initiated instruction) event. It helps identify points that players feel are too difficult.

731
This event occurs when the player clicks on the help button. It contains information about the state of the

In [82]:
specs.loc[specs.event_id=='532a2afb']['info'].item()

/Users/loftis/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: `item` has been deprecated and will be removed in a future version
  """Entry point for launching an IPython kernel.


'The start round event is triggered at the start of a round when the player is prompted to place the cauldrons and when the player is prompted to select a certain cauldron. This event provides information about the game characteristics of the round (i.e. resources, objectives, setup). It is used in calculating things like time spent in a round (for speed and accuracy), attempts at solving a round, and the number of rounds the player has visited (exposures).'

In [135]:
specs.loc[specs.event_id=='90d848e0','info'].item()

/Users/loftis/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: `item` has been deprecated and will be removed in a future version
  """Entry point for launching an IPython kernel.


"The start game event is triggered at the very beginning of the level (after the game finishes loading, don't wait for intro movie to finish). This is used to compute things like time spent in game."

In [13]:
# find first assesment for each label in train_set_labels
min_time_stamp = (train_data.loc[train_data.game_session.isin(train_labels.game_session)]
                  .groupby('game_session')
                  .agg({'timestamp':'min'})
                  .reset_index())
min_time_stamp.head()

game_session                        timestamp
0  00097cda27afb726 2019-09-15 16:36:41.450000+00:00
1  000f68cff32664ef 2019-10-12 23:07:25.711000+00:00
2  0014403daadf67aa 2019-08-03 02:43:31.845000+00:00
3  0014daa1d3e26eb2 2019-09-24 22:28:39.513000+00:00
4  001c49e9e9968dbe 2019-07-26 15:50:36.433000+00:00

In [33]:
merged_ = (train_data.loc[train_data.game_session.isin(min_time_stamp.game_session)]
                     .merge(min_time_stamp,on='game_session'))
merged_['first_event_of_assessment'] = False
merged_.loc[merged_.timestamp_x == merged_.timestamp_y,'first_event_of_assessment'] = 1
merged_ = merged_.loc[:,['id','first_event_of_assessment']]

In [34]:
train_data_aug = train_data.merge(merged_,on='id',how='left')

In [35]:
train_data_aug['first_event_of_assessment'].fillna(0,inplace=True)
train_data_aug.head()

event_id      game_session                        timestamp  \
0  27253bdc  45bb1e1b6b50c07b 2019-09-06 17:53:46.937000+00:00   
1  27253bdc  17eeb7f223665f53 2019-09-06 17:54:17.519000+00:00   
2  77261ab5  0848ef14a8dc6892 2019-09-06 17:54:56.302000+00:00   
3  b2dba42b  0848ef14a8dc6892 2019-09-06 17:54:56.387000+00:00   
4  1bb5fbdb  0848ef14a8dc6892 2019-09-06 17:55:03.253000+00:00   

                                          event_data installation_id  \
0             {"event_code": 2000, "event_count": 1}        0001e90f   
1             {"event_code": 2000, "event_count": 1}        0001e90f   
2  {"version":"1.0","event_count":1,"game_time":0...        0001e90f   
3  {"description":"Let's build a sandcastle! Firs...        0001e90f   
4  {"description":"Let's build a sandcastle! Firs...        0001e90f   

   event_count  event_code  game_time                          title  \
0            1        2000          0        Welcome to Lost Lagoon!   
1            1        2000          0           Magma Peak - Level 1   
2            1        2000          0  Sandcastle Builder (Activity)   
3            2        3010         53  Sandcastle Builder (Activity)   
4            3        3110       6972  Sandcastle Builder (Activity)   

       type      world  id first_event_of_assessment  
0      Clip       NONE   0                         0  
1      Clip  MAGMAPEAK   1                         0  
2  Activity  MAGMAPEAK   2                         0  
3  Activity  MAGMAPEAK   3                         0  
4  Activity  MAGMAPEAK   4                         0

In [46]:
train_data_big = train_data_aug.merge(train_labels,on='installation_id')
train_data_big

event_id    game_session_x                        timestamp  \
0         27253bdc  34ba1a28d02ba8ba 2019-08-06 04:57:18.904000+00:00   
1         27253bdc  34ba1a28d02ba8ba 2019-08-06 04:57:18.904000+00:00   
2         27253bdc  34ba1a28d02ba8ba 2019-08-06 04:57:18.904000+00:00   
3         27253bdc  34ba1a28d02ba8ba 2019-08-06 04:57:18.904000+00:00   
4         27253bdc  34ba1a28d02ba8ba 2019-08-06 04:57:18.904000+00:00   
...            ...               ...                              ...   
91821964  b5053438  5448d652309a6324 2019-09-22 02:07:28.311000+00:00   
91821965  b5053438  5448d652309a6324 2019-09-22 02:07:28.311000+00:00   
91821966  2b058fe3  5448d652309a6324 2019-09-22 02:07:31.963000+00:00   
91821967  2b058fe3  5448d652309a6324 2019-09-22 02:07:31.963000+00:00   
91821968  2b058fe3  5448d652309a6324 2019-09-22 02:07:31.963000+00:00   

                                                 event_data installation_id  \
0                    {"event_code": 2000, "event_count": 1}        0006a69f   
1                    {"event_code": 2000, "event_count": 1}        0006a69f   
2                    {"event_code": 2000, "event_count": 1}        0006a69f   
3                    {"event_code": 2000, "event_count": 1}        0006a69f   
4                    {"event_code": 2000, "event_count": 1}        0006a69f   
...                                                     ...             ...   
91821964  {"description":"Awesome.","identifier":"Dot_Aw...        ffeb0b1b   
91821965  {"description":"Awesome.","identifier":"Dot_Aw...        ffeb0b1b   
91821966  {"session_duration":71511,"event_count":61,"ga...        ffeb0b1b   
91821967  {"session_duration":71511,"event_count":61,"ga...        ffeb0b1b   
91821968  {"session_duration":71511,"event_count":61,"ga...        ffeb0b1b   

          event_count  event_code  game_time                       title_x  \
0                   1        2000          0       Welcome to Lost Lagoon!   
1                   1        2000          0       Welcome to Lost Lagoon!   
2                   1        2000          0       Welcome to Lost Lagoon!   
3                   1        2000          0       Welcome to Lost Lagoon!   
4                   1        2000          0       Welcome to Lost Lagoon!   
...               ...         ...        ...                           ...   
91821964           60        3121      67847  Cauldron Filler (Assessment)   
91821965           60        3121      67847  Cauldron Filler (Assessment)   
91821966           61        2010      71511  Cauldron Filler (Assessment)   
91821967           61        2010      71511  Cauldron Filler (Assessment)   
91821968           61        2010      71511  Cauldron Filler (Assessment)   

                type      world        id first_event_of_assessment  \
0               Clip       NONE      1538                         0   
1               Clip       NONE      1538                         0   
2               Clip       NONE      1538                         0   
3               Clip       NONE      1538                         0   
4               Clip       NONE      1538                         0   
...              ...        ...       ...                       ...   
91821964  Assessment  MAGMAPEAK  11337824                     False   
91821965  Assessment  MAGMAPEAK  11337824                     False   
91821966  Assessment  MAGMAPEAK  11337825                     False   
91821967  Assessment  MAGMAPEAK  11337825                     False   
91821968  Assessment  MAGMAPEAK  11337825                     False   

            game_session_y                       title_y  num_correct  \
0         6bdf9623adc94d89  Mushroom Sorter (Assessment)            1   
1         77b8ee947eb84b4e    Bird Measurer (Assessment)            0   
2         901acc108f55a5a1  Mushroom Sorter (Assessment)            1   
3         9501794defd84e4d  Mushroom Sorter (Assessment)            1   
4         a9ef3ecb3d1acc6a   

In [49]:
train_data_big[['timestamp','installation_id','id','game_session_y']].head()

timestamp installation_id    id    game_session_y
0 2019-08-06 04:57:18.904000+00:00        0006a69f  1538  6bdf9623adc94d89
1 2019-08-06 04:57:18.904000+00:00        0006a69f  1538  77b8ee947eb84b4e
2 2019-08-06 04:57:18.904000+00:00        0006a69f  1538  901acc108f55a5a1
3 2019-08-06 04:57:18.904000+00:00        0006a69f  1538  9501794defd84e4d
4 2019-08-06 04:57:18.904000+00:00        0006a69f  1538  a9ef3ecb3d1acc6a

In [50]:
train_data_big.shape

(91821969, 19)

In [26]:
train_labels.head()

game_session installation_id                         title  \
0  6bdf9623adc94d89        0006a69f  Mushroom Sorter (Assessment)   
1  77b8ee947eb84b4e        0006a69f    Bird Measurer (Assessment)   
2  901acc108f55a5a1        0006a69f  Mushroom Sorter (Assessment)   
3  9501794defd84e4d        0006a69f  Mushroom Sorter (Assessment)   
4  a9ef3ecb3d1acc6a        0006a69f    Bird Measurer (Assessment)   

   num_correct  num_incorrect  accuracy  accuracy_group  
0            1              0       1.0               3  
1            0             11       0.0               0  
2            1              0       1.0               3  
3            1              1       0.5               2  
4            1              0       1.0               3

In [148]:
train_data.loc[train_data.game_session.isin(train_labels.game_session)]

event_id      game_session                        timestamp  \
2185      3bfd1a65  901acc108f55a5a1 2019-08-06 05:22:01.344000+00:00   
2186      db02c830  901acc108f55a5a1 2019-08-06 05:22:01.400000+00:00   
2187      a1e4395d  901acc108f55a5a1 2019-08-06 05:22:01.403000+00:00   
2188      a52b92d5  901acc108f55a5a1 2019-08-06 05:22:05.242000+00:00   
2189      a1e4395d  901acc108f55a5a1 2019-08-06 05:22:05.244000+00:00   
...            ...               ...                              ...   
11337821  28520915  5448d652309a6324 2019-09-22 02:07:27.562000+00:00   
11337822  91561152  5448d652309a6324 2019-09-22 02:07:27.562000+00:00   
11337823  d3268efa  5448d652309a6324 2019-09-22 02:07:27.566000+00:00   
11337824  b5053438  5448d652309a6324 2019-09-22 02:07:28.311000+00:00   
11337825  2b058fe3  5448d652309a6324 2019-09-22 02:07:31.963000+00:00   

                                                 event_data installation_id  \
2185      {"version":"1.0","event_count":1,"game_time":0...        0006a69f   
2186      {"event_count":2,"game_time":37,"event_code":2...        0006a69f   
2187      {"description":"Pull three mushrooms out of th...        0006a69f   
2188      {"description":"Pull three mushrooms out of th...        0006a69f   
2189      {"description":"To pick a mushroom, pull it ou...        0006a69f   
...                                                     ...             ...   
11337821  {"misses":1,"prompt":"holds least","mode":"sel...        ffeb0b1b   
11337822  {"bucket":1,"buckets_placed":[3,1,2],"target_b...        ffeb0b1b   
11337823  {"description":"Awesome.","identifier":"Dot_Aw...        ffeb0b1b   
11337824  {"description":"Awesome.","identifier":"Dot_Aw...        ffeb0b1b   
11337825  {"session_duration":71511,"event_count":61,"ga...        ffeb0b1b   

          event_count  event_code  game_time                         title  \
2185                1        2000          0  Mushroom Sorter (Assessment)   
2186                2        2025         37  Mushroom Sorter (Assessment)   
2187                3        3010         37  Mushroom Sorter (Assessment)   
2188                4        3110       3901  Mushroom Sorter (Assessment)   
2189                5        3010       3901  Mushroom Sorter (Assessment)   
...               ...         ...        ...                           ...   
11337821           58        2030      67094  Cauldron Filler (Assessment)   
11337822           57        4025      67094  Cauldron Filler (Assessment)   
11337823           59        3021      67094  Cauldron Filler (Assessment)   
11337824           60        3121      67847  Cauldron Filler (Assessment)   
11337825           61        2010      71511  Cauldron Filler (Assessment)   

                type        world        id  
2185      Assessment  TREETOPCITY      2185  
2186      Assessment  TREETOPCITY      2186  
2187      Assessment  TREETOPCITY      2187  
2188      Assessment  TREETOPCITY      2188  
2189      Assessment  TREETOPCITY      2189  
...              ...          ...       ...  
11337821  Assessment    MAGMAPEAK  11337821  
11337822  Assessment    MAGMAPEAK  11337822  
11337823  Assessment    MAGMAPEAK  11337823  
11337824  Assessment    MAGMAPEAK  11337824  
11337825  Assessment    MAGMAPEAK  11337825  

[865447 rows x 12 columns]

In [142]:
test_data.sample(5)

event_id      game_session                 timestamp  \
639286   56bcd38d  fa9f4e65edb7efac  2019-08-24T18:57:38.779Z   
675077   76babcde  df14150f0f1d862e  2019-07-31T14:45:35.921Z   
781464   acf5c23f  9a299ad5105c07f0  2019-07-25T13:50:00.938Z   
990721   16dffff1  7645e395f6994a74  2019-08-16T12:46:12.346Z   
1118483  5e812b27  773d3a45bbda407a  2019-08-24T17:22:38.183Z   

                                                event_data installation_id  \
639286   {"object":"chicken","layout":{"left":{"chicken...        8cd5bc7c   
675077   {"coordinates":{"x":854,"y":427,"stage_width":...        9069db29   
781464   {"coordinates":{"x":176,"y":680,"stage_width":...        a41a15f2   
990721   {"duration":49987,"misses":2,"round":2,"event_...        d719a0ac   
1118483  {"size":0,"coordinates":{"x":782,"y":207,"stag...        f86a6ed4   

         event_count  event_code  game_time                          title  \
639286           221        4030     194105    Chicken Balancer (Activity)   
675077           160        4070     296389                      Dino Dive   
781464            98        4070     103798     Cart Balancer (Assessment)   
990721            59        2030      82064                     Dino Drink   
1118483          114        4030     125906  Sandcastle Builder (Activity)   

               type         world  
639286     Activity  CRYSTALCAVES  
675077         Game     MAGMAPEAK  
781464   Assessment  CRYSTALCAVES  
990721         Game     MAGMAPEAK  
1118483    Activity     MAGMAPEAK

In [147]:
test_data.loc[test_data.installation_id=='d719a0ac']

event_id      game_session                 timestamp  \
990624  27253bdc  466d9d2b57d3810e  2019-08-07T14:24:15.838Z   
990625  27253bdc  94fa0866c6638da9  2019-08-07T14:24:44.398Z   
990626  77261ab5  10893d61b3b01f49  2019-08-07T14:25:14.115Z   
990627  b2dba42b  10893d61b3b01f49  2019-08-07T14:25:14.253Z   
990628  1325467d  10893d61b3b01f49  2019-08-07T14:25:18.828Z   
...          ...               ...                       ...   
990752  d3268efa  8c6a75dfe2afb3b6  2019-08-16T12:47:01.841Z   
990753  b5053438  8c6a75dfe2afb3b6  2019-08-16T12:47:02.578Z   
990754  2b058fe3  8c6a75dfe2afb3b6  2019-08-16T12:47:06.226Z   
990755  27253bdc  72d2f1fd02fbe18e  2019-08-16T12:47:26.459Z   
990756  3bfd1a65  50a0a2f52eec28a8  2019-08-16T12:47:58.221Z   

                                               event_data installation_id  \
990624             {"event_code": 2000, "event_count": 1}        d719a0ac   
990625             {"event_code": 2000, "event_count": 1}        d719a0ac   
990626  {"version":"1.0","event_count":1,"game_time":0...        d719a0ac   
990627  {"description":"Let's build a sandcastle! Firs...        d719a0ac   
990628  {"coordinates":{"x":830,"y":633,"stage_width":...        d719a0ac   
...                                                   ...             ...   
990752  {"description":"Awesome.","identifier":"Dot_Aw...        d719a0ac   
990753  {"description":"Awesome.","identifier":"Dot_Aw...        d719a0ac   
990754  {"session_duration":32459,"event_count":32,"ga...        d719a0ac   
990755             {"event_code": 2000, "event_count": 1}        d719a0ac   
990756  {"version":"1.0","event_count":1,"game_time":0...        d719a0ac   

        event_count  event_code  game_time                          title  \
990624            1        2000          0        Welcome to Lost Lagoon!   
990625            1        2000          0           Magma Peak - Level 1   
990626            1        2000          0  Sandcastle Builder (Activity)   
990627            2        3010         60  Sandcastle Builder (Activity)   
990628            3        4070       4737  Sandcastle Builder (Activity)   
...             ...         ...        ...                            ...   
990752           30        3021      28060   Cauldron Filler (Assessment)   
990753           31        3121      28809   Cauldron Filler (Assessment)   
990754           32        2010      32459   Cauldron Filler (Assessment)   
990755            1        2000          0        Tree Top City - Level 1   
990756            1        2000          0   Mushroom Sorter (Assessment)   

              type        world  
990624        Clip         NONE  
990625        Clip    MAGMAPEAK  
990626    Activity    MAGMAPEAK  
990627    Activity    MAGMAPEAK  
990628    Activity    MAGMAPEAK  
...            ...          ...  
990752  Assessment    MAGMAPEAK  
990753  Assessment    MAGMAPEAK  
990754  Assessment    MAGMAPEAK  
990755        Clip  TREETOPCITY  
990756  Assessment  TREETOPCITY  

[133 rows x 11 columns]

In [121]:
train_labels.game_session.nunique()

17690

In [122]:
train_labels.shape

(17690, 7)

In [96]:
train_data_labels_merged

event_id      game_session                 timestamp  \
0         27253bdc  45bb1e1b6b50c07b  2019-09-06T17:53:46.937Z   
1         27253bdc  17eeb7f223665f53  2019-09-06T17:54:17.519Z   
2         77261ab5  0848ef14a8dc6892  2019-09-06T17:54:56.302Z   
3         b2dba42b  0848ef14a8dc6892  2019-09-06T17:54:56.387Z   
4         1bb5fbdb  0848ef14a8dc6892  2019-09-06T17:55:03.253Z   
...            ...               ...                       ...   
11341037  ab3136ba  c09b94eebfdf50a6  2019-10-10T15:11:15.584Z   
11341038  27253bdc  6ad8ab25003ef1b0  2019-10-10T15:12:17.364Z   
11341039  27253bdc  71904dca23b982c6  2019-10-10T15:13:04.713Z   
11341040  27253bdc  2267b18f069e875a  2019-10-10T15:13:54.545Z   
11341041  27253bdc  f31950f549d7c242  2019-10-10T15:15:12.483Z   

                                                 event_data installation_id_x  \
0                    {"event_code": 2000, "event_count": 1}          0001e90f   
1                    {"event_code": 2000, "event_count": 1}          0001e90f   
2         {"version":"1.0","event_count":1,"game_time":0...          0001e90f   
3         {"description":"Let's build a sandcastle! Firs...          0001e90f   
4         {"description":"Let's build a sandcastle! Firs...          0001e90f   
...                                                     ...               ...   
11341037  {"description":"Tap another dinosaur.","identi...          fffc0583   
11341038             {"event_code": 2000, "event_count": 1}          fffc0583   
11341039             {"event_code": 2000, "event_count": 1}          fffc0583   
11341040             {"event_code": 2000, "event_count": 1}          fffc0583   
11341041             {"event_code": 2000, "event_count": 1}          fffc0583   

          event_count  event_code  game_time                        title_x  \
0                   1        2000          0        Welcome to Lost Lagoon!   
1                   1        2000          0           Magma Peak - Level 1   
2                   1        2000          0  Sandcastle Builder (Activity)   
3                   2        3010         53  Sandcastle Builder (Activity)   
4                   3        3110       6972  Sandcastle Builder (Activity)   
...               ...         ...        ...                            ...   
11341037           66        3110     118615                      Dino Dive   
11341038            1        2000          0        Crystal Caves - Level 2   
11341039            1        2000          0        Crystal Caves - Level 3   
11341040            1        2000          0        Crystal Caves - Level 3   
11341041            1        2000          0        Tree Top City - Level 2   

              type         world installation_id_y title_y  num_correct  \
0             Clip          NONE               NaN     NaN          NaN   
1             Clip     MAGMAPEAK               NaN     NaN          NaN   
2         Activity     MAGMAPEAK               NaN     NaN          NaN   
3         Activity     MAGMAPEAK               NaN     NaN          NaN   
4         Activity     MAGMAPEAK               NaN     NaN          NaN   
...            ...           ...               ...     ...          ...   
11341037      Game     MAGMAPEAK               NaN     NaN          NaN   
11341038      Clip  CRYSTALCAVES               NaN     NaN          NaN   
11341039      Clip  CRYSTALCAVES               NaN     NaN          NaN   
11341040      Clip  CRYSTALCAVES               NaN     NaN          NaN   
11341041      Clip   TREETOPCITY               NaN     NaN          NaN   

          num_incorrect  accuracy  accuracy_group  
0                   NaN       NaN             NaN  
1                   NaN       NaN             NaN  
2                   NaN       NaN             NaN  
3                   NaN       NaN             NaN  
4                   NaN       NaN             NaN  
...                 ...       ...             ...  
11341037            NaN       

In [85]:
train_labels.head()

game_session installation_id                         title  \
0  6bdf9623adc94d89        0006a69f  Mushroom Sorter (Assessment)   
1  77b8ee947eb84b4e        0006a69f    Bird Measurer (Assessment)   
2  901acc108f55a5a1        0006a69f  Mushroom Sorter (Assessment)   
3  9501794defd84e4d        0006a69f  Mushroom Sorter (Assessment)   
4  a9ef3ecb3d1acc6a        0006a69f    Bird Measurer (Assessment)   

   num_correct  num_incorrect  accuracy  accuracy_group  
0            1              0       1.0               3  
1            0             11       0.0               0  
2            1              0       1.0               3  
3            1              1       0.5               2  
4            1              0       1.0               3

## Feature Engineering:

- As a first pass at engineering some features we created for each installation id a column for the number of each event ID that a given installation id performed.
- number of game sessions

In [ ]:
# compute number of event_ids per installation_id
